internal/testintegration/setup.go

```diff
@@ -110,85 +106,147 @@ func EnvTestSetup() (func(), error) {
-		cancel()
-		err := testEnv.Stop()
-		if err != nil {
-			Log.Error(err, "unable to stop envtest environment %v")
-		}
-	}
-
+	// Start the testenv
 	cfg, err := testEnv.Start()
+	mh := &EnvTestHarness{
+		testEnvCtx:    ctx,
+		testEnv:       testEnv,
+		testEnvCancel: cancel,
+		cfg:           cfg,
+	}
 	if err != nil {
-		return nil, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return mh, fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
-	s := scheme.Scheme
-	controller.InitScheme(s)
-
-	Client, err = client.New(cfg, client.Options{Scheme: s})
+	// Initialize rest client configuration
+	mh.s = scheme.Scheme
+	controller.InitScheme(mh.s)
+	cl, err := client.New(cfg, client.Options{Scheme: mh.s})
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return mh, fmt.Errorf("unable to to create client %v", err)
 	}
-	if Client == nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+	if cl == nil {
+		return mh, fmt.Errorf("client not created")
 	}
+	mh.Client = cl
 
-	mgrCancelFunc, err := startManager(ctx, cfg, s, workload.DefaultProxyImage)
+	// Start the controller-runtime manager
+	err = mh.StartMgr(workload.DefaultProxyImage)
 	if err != nil {
-		return teardownFunc, fmt.Errorf("unable to start kuberenetes envtest %v", err)
+		return mh, fmt.Errorf("unable to start kuberenetes envtest %v", err)
 	}
 
-	RestartManager = func(defaultProxyImage string) error {
-		mgrCancelFunc()
-		mgrCancelFunc, err = startManager(ctx, cfg, s, defaultProxyImage)
-		return err
+	return mh, nil
+}
+
+// EnvTestHarness enables integration tests to control the lifecycle of the
+// operator's controller-manager.
+type EnvTestHarness struct {
+
+	// Mgr is the manager
+	Mgr ctrl.Manager
+
+	// Client is the kubernetes client.
+	Client client.Client
+
+	// The actual EnvTest environment
+	testEnv *envtest.Environment
+
+	// testEnvCancel is the context cancel function for the testEnv
+	testEnvCancel context.CancelFunc
+
+	// cancel is the context cancel function for the manager
+	cancel context.CancelFunc
+
+	// stopped channel is closed when the manager actually stops
+	stopped chan int
+
+	// ctx is the manager's context
+	ctx context.Context
+
+	// cfg is the client configuration from envtest
+	cfg *rest.Config
+
+	// s is the client scheme
+	s          *runtime.Scheme
+	testEnvCtx context.Context
+}
+
+// Teardown closes the TestEnv environment at the end of the testcase.
+func (h *EnvTestHarness) Teardown() {
+	if h.testEnvCancel != nil {
+		h.testEnvCancel()
+	}
+	if h.testEnv != nil {
+		err := h.testEnv.Stop()
+		if err != nil {
+			Log.Error(err, "unable to stop envtest environment %v")
+		}
 	}
+}
 
-	return teardownFunc, nil
+// StopMgr stops the controller manager and waits for it to exit, returning an
+// error if the controller manager does not stop within 1 minute.
+func (h *EnvTestHarness) StopMgr() error {
+	if h.cancel != nil {
```

Comment:
When is `cancel` nil? Could we avoid this nil check with better initialization?

Reply:
Unfortunately no. If EnvTestHarness.StopMgr() is called when the manager is already stopping, h.cancel will be null.

Reply:
Could we just initialize `h.cancel` to be a no-op function in the initializer and skip the nil checking?

Reply:
Fixed.